packages/jupyter-ai/src/chat_handler.ts

```diff
@@ -164,6 +140,41 @@ export class ChatHandler implements IDisposable {
     (value: AiService.AgentChatMessage) => void
   > = {};
 
+  private _onClose(reject: any) {
+    reject(new Error("Chat UI websocket disconnected"))
+    console.error("Chat UI websocket disconnected")
+    setTimeout(async () => await this._initialize(), 500);
+  }
+
+  private _initialize(): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      if (this.isDisposed) {
+        return;
+      }
+      console.log("Creating a new websocket connection...");
+      const { token, WebSocket, wsUrl } = this.serverSettings;
+      const url =
+        URLExt.join(wsUrl, CHAT_SERVICE_URL) +
+        (token ? `?token=${encodeURIComponent(token)}` : '');
+      
+        const socket = (this._socket = new WebSocket(url));
+        socket.onclose = () => this._onClose(reject);
+        socket.onerror = (e) => reject(e);
+        socket.onmessage = msg =>
+          msg.data && this._onMessage(JSON.parse(msg.data));
+        const listenForConnection = (message: AiService.Message) => {
+          if (message.type !== 'connection') {
+            return;
+          }
+          this.id = message.client_id;
+          resolve();
+          this.removeListener(listenForConnection);
+        };
+  
+        this.addListener(listenForConnection);
```

Comment:
Should we have a timeout if no connection is made, so we can try again?

Reply:
I am not sure what a reasonable timeout should be that will work across different infrastructure setups. I believe browser will take care of the timeout, and the WebSocket will error out and close connection after that timeout.